File: src/features/describe/describe.service.ts
import { buildPrompt } from './describe.prompt'
import { resolveTargets } from '../../infrastructure/git/git-context-service'
import { parseWorkItemId } from '../../shared/validation/work-item'
import type { DescriptionGenerator } from '../../infrastructure/ai/description-generator'
import type { PullRequestPublisher } from '../../infrastructure/azure/pull-request-publisher'
import type { GitContextReader } from '../../infrastructure/git/git-context-reader'
import type { ConfigService } from '../../infrastructure/config/config-service'
import type { CliOptions } from '../../app/cli.models'
import type { PrDescription, ProviderName } from '../../infrastructure/ai/ai.models'
import type { DescribePreparation, GeneratedDescription } from './describe.models'

export class DescribeService {
  constructor(
    private readonly config: ConfigService,
    private readonly git: GitContextReader,
    private readonly generator: DescriptionGenerator,
    private readonly publisher: PullRequestPublisher
  ) {}

  prepare(options: CliOptions, interactive: boolean): DescribePreparation {
    const config = this.config.load(options)
    const context = this.git.collect(options.source)
    if (options.targets.includes('sprint') && !context.sprintBranch) {
      throw new Error('Target sprint solicitado, mas nenhuma branch sprint/<número> foi encontrada.')
    }
    const workItemId = options.workItem ?? context.workItemId
    if (workItemId) parseWorkItemId(workItemId, 'Work Item')
    const targets = resolveTargets(context, options.targets)
    if (targets.length === 0) throw new Error('Nenhum target disponível.')
    return {
      config,
      context,
      targets,
      workItemId,
      system: config.template,
      prompt: buildPrompt(context, targets, workItemId ?? ''),
      interactive
    }
  }

  validateCreation(preparation: DescribePreparation, requested: boolean): void {
    if (requested && !preparation.interactive) {
      throw new Error('--create requer terminal interativo para confirmar a descrição e os revisores.')
    }
    if (requested && !preparation.context.isAzureDevOps) {
      throw new Error('--create requer um remote Git do Azure DevOps.')
    }
    if (requested && !preparation.config.azurePat.trim()) {
      throw new Error('--create requer AZURE_PAT ou AZURE_DEVOPS_PAT configurado.')
    }
  }

  generate(
    preparation: DescribePreparation,
    report: (provider: ProviderName, model: string) => void
  ): Promise<GeneratedDescription> {
    return this.generator.generate({
      config: preparation.config,
      system: preparation.system,
      prompt: preparation.prompt,
      branch: preparation.context.branch,
      report
    })
  }

  publish(
    preparation: DescribePreparation,
    description: PrDescription,
    reviewerForTarget?: (target: string) => string
  ) {
    return this.publisher.publish(
      preparation.config,
      preparation.context,
      preparation.targets,
      {
        title: description.title,
        description: description.body,
        workItemRefs: preparation.workItemId ? [{ id: preparation.workItemId }] : undefined
      },
      reviewerForTarget
    )
  }
}

File: lib/src/features/describe/describe_service.dart
import 'package:better_effect/better_effect.dart';

import '../../application/ai/description_generator.dart';
import '../../application/ai/description_models.dart';
import '../../app/app_effect.dart';
import '../../app/app_failure.dart';
import '../../app/cli_options.dart';
import '../../application/config/config_service.dart';
import '../../application/change_context/change_context_reader.dart';
import '../../application/change_context/merged_pull_request_lookup.dart';
import '../../domain/change_context.dart';
import 'describe_models.dart';
import 'describe_prompt.dart';

abstract interface class DescribeService {
  AppEffect<DescribePreparation> prepare(CliOptions options, bool interactive);

  AppEffect<Unit> validateCreation(
    DescribePreparation preparation,
    bool requested,
  );

  AppEffect<GeneratedDescription> generate(
    DescribePreparation preparation, {
    DescriptionReporter? report,
  });
}

final class DescribeFailure extends AppFailure {
  const DescribeFailure(String message) : super(message, 1);
}

final class DescribeServiceLive implements DescribeService {
  const DescribeServiceLive();

  @override
  AppEffect<DescribePreparation> prepare(
    CliOptions options,
    bool interactive,
  ) => .result((use) async {
    final config = await use.unwrap(use<ConfigService>().load(options));
    final initialContext = await use.unwrap(
      use<ChangeContextReader>().collect(options.source),
    );
    final usesDefaultTargets = options.targets.isEmpty;
    if ((usesDefaultTargets || options.targets.contains('sprint')) &&
        initialContext.sprintBranch.isEmpty) {
      use.fail(
        const DescribeFailure(
          'Target sprint solicitado, mas nenhuma branch sprint/<número> foi encontrada.',
        ),
      );
    }
    final targets = resolveTargets(initialContext, options.targets);
    if (targets.isEmpty) {
      use.fail(const DescribeFailure('Nenhum target disponível.'));
    }
    final mergedResult = await use.unwrap(
      use<MergedPullRequestLookup>().findLatest(
        remote: initialContext.remote,
        sourceBranch: initialContext.branch,
        targetBranches: targets,
      ),
    );
    final mergedPullRequest = mergedResult.value;
    final context = mergedPullRequest == null
        ? initialContext
        : await use.unwrap(
            use<ChangeContextReader>().collect(
              options.source,
              mergedPullRequest.sourceCommit,
            ),
          );
    final workItemId =
        options.workItem?.value ??
        (context.workItemId.isEmpty ? null : context.workItemId);
    if (workItemId != null) {
      await use.result(_validateWorkItem(workItemId, 'Work Item'));
    }
    return DescribePreparation(
      config: config,
      context: context,
      targets: targets,
      workItemId: workItemId,
      system: config.template,
      prompt: buildPrompt(context, targets, workItemId ?? ''),
      interactive: interactive,
    );
  });

  @override
  AppEffect<Unit> validateCreation(
    DescribePreparation preparation,
    bool requested,
  ) => .result((use) {
    if (requested && !preparation.interactive) {
      use.fail(
        const DescribeFailure(
          '--create requer terminal interativo para confirmar a descrição e os revisores.',
        ),
      );
    }
    if (requested && preparation.context.remote == null) {
      use.fail(
        const DescribeFailure('--create requer um remote Git do Azure DevOps.'),
      );
    }
    if (requested && preparation.config.azurePat.trim().isEmpty) {
      use.fail(
        const DescribeFailure(
          '--create requer AZURE_PAT ou AZURE_DEVOPS_PAT configurado.',
        ),
      );
    }
    return unit;
  });

  @override
  AppEffect<GeneratedDescription> generate(
    DescribePreparation preparation, {
    DescriptionReporter? report,
  }) => .result((use) async {
    final generator = use<DescriptionGenerator>();
    return use.unwrap(
      generator.generate(
        config: preparation.config,
        system: preparation.system,
        prompt: preparation.prompt,
        branch: preparation.context.branch,
        report: report,
      ),
    );
  });
}

ResultDart<String, DescribeFailure> _validateWorkItem(
  String value,
  String label,
) {
  final id = value.trim();
  if (!RegExp(r'^\d+$').hasMatch(id)) {
    return Failure<String, DescribeFailure>(
      DescribeFailure('$label inválido: Use um ID numérico.'),
    );
  }
  final number = int.tryParse(id);
  if (number == null || number <= 0) {
    return Failure<String, DescribeFailure>(
      DescribeFailure('$label inválido: Use um ID positivo.'),
    );
  }
  return Success<String, DescribeFailure>(id);
}
